src/box/txn_limbo.h

```diff
@@ -326,6 +327,7 @@ txn_limbo_wait_complete(struct txn_limbo *limbo, struct txn_limbo_entry *entry);
 static inline void
 txn_limbo_begin(struct txn_limbo *limbo)
 {
+	ERROR_INJECT_YIELD(ERRINJ_TXN_LIMBO_BEGIN_DELAY);
```

Comment:
We again stumble upon a new error injection which could be avoided. I understand that it looks really simple to just add one per new bugfix, but it really makes it less likely that we test something real.

Here in terms of yields all is fine, `latch_lock()` below can yield. But it yields in specific cases. In particular, if there is another `txn_limbo_begin()` called somewhere. Which means not only the latch but generally a bit broader part of the limbo state can be affected.

I wanted to make an example with `CONFIRM`, but I see `txn_limbo_req_prepare()` doesn't do anything with `CONFIRM`, so bad example. But that can change in the future and the test can become obsolete, incomplete.

Anyway, it seems you could either use the advice from Sergey, or take the original test case from the ticket (there is a huge test file, but I suppose the crash is in one specific case in that file).

Reply:
I've did a lot of experiments, trying to avoid adding new error injections. box.ctl.promote() goes kinda like:
1) Increment volatile_term
2) Wait until volatile_term is written to WAL
2.1) Check that no one else promoted/demoted during WAL write, raise ER_INTERFERING_PROMOTE otherwise
3) Elections (if `box.cfg.election_mode` is not `off`)
4) Call box_issue_promote, where
4.1) Assert that volatile_term is written to disk (`raft->volatile_term == raft->term`)
4.2) Try to lock `promote_latch` by calling `txn_limbo_begin`
4.3) Assert raft term wasn't changed while locking the latch
4.4) Assert no one else promoted/demoted while locking the latch

To test the code in question, we need to receive new term, or promote in current term, while 4.2 is waiting.
I've tried locking the latch by issuing promote on another server with the following test:
```lua
g.test_txn_limbo_begin_interfering_promote = function(g)
    wal_delay_start(g.server_1)
    local f = promote_start(g.server_1)

    promote(g.server_2)
    g.server_3:wait_election_term(g.server_2:election_term())

    promote(g.server_3)
    g.server_2:wait_election_term(g.server_3:election_term())

    wal_delay_end(g.server_1)
    local ok, err = fiber_join(g.server_1, f)

    print(ok, err)
end
```
Desired output is "false   Interfering elections started". But this is not always the case, half of the times I get "false   Instance with replica id 3 was promoted first", which could look ok at first, but when i go through gcov output, I see that in second case its box_trigger_elections() that fails, not box_issue_promote(). I've tried to stop server_1 not during writing new term to WAL, but during the elections process (by setting it's election_mode to `manual` while on others it's set to `off`) and releasing it after lock was obtained by neighbour by setting `box.cfg{replication_synchro_quorum = 1}`. It didn't work too, because box_run_elections fails.

While trying to get this test to work without new error injection, I've found a test, that triggers `assert(raft->volatile_term == term)` in box_issue_promote. I've removed this assert in favor of checking it, and throwing error if it's not equal.

IMO this error injection is fine, because currently this latch can also be locked by applier_synchro_filter_tx() in applier thread , while processing replication almost at any time.

Reply:
> IMO this error injection is fine

This whole response above is crying that it is not fine though :D - you found and fixed an assertion, which wouldn't be triggered by the hacky injection.

You also say that a new test is failing only half times - it would be fine. Not each test needs to fail without the patch 100% runs. If it would fail even 10% runs, it would be already tolerable. It would catch the issue anyway if it would return.

> because currently this latch can also be locked by applier_synchro_filter_tx()

This brings us to my initial point - it could be locked here too, but by adding the injection you do not test that case. You test something impossible instead.

Nonetheless I am fine with it, lets keep this injection if you want.